### PR TITLE
feat (ui): manage permissions on environments

### PR DESCRIPTION
**Pull request by sguiheux**

### CODE REVIEW

File: ui/src/app/shared/permission/environment/form/permission.env.form.component.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            let newGroups: Array<Group> = this.selectedGroups.map(ids => {
+               let isHere = env.groups.find(gp => gp.group.id === Number(ids));
+               if (isHere) {
+                   return;
+               }
+               return this.groups.find(g => g.id === Number(ids));
+            });
+
+            let gps = new Array<GroupPermission>();
+            newGroups.forEach(g => {
```
Comment (bnjjj) [inline ui/src/app/shared/permission/environment/form/permission.env.form.component.ts]: `.map`more explicit in this case

File: ui/src/app/views/project/show/project.component.ts [hunk 2]
```diff
@@ -112,6 +121,43 @@ export class ProjectShowComponent implements OnInit, OnDestroy {
...
+            this.permEnvGroupWarningModal.show(event);
+        } else {
+            if (!env) {
+                env = event.env;
+            }
+            switch (event.type) {
+                case 'update':
+                    this._projectStore.updateEnvironmentPermission(this.project.key, env.name, event.gp).subscribe(() => {
+                        this._toast.success('', this._translate.instant('permission_updated'));
+                    });
```
Comment (bnjjj) [inline ui/src/app/views/project/show/project.component.ts]: and if it fails ?

File: ui/src/app/views/project/show/project.component.ts [hunk 3]
```diff
@@ -112,6 +121,43 @@ export class ProjectShowComponent implements OnInit, OnDestroy {
...
+                    break;
+                case 'delete':
+                    this._projectStore.removeEnvironmentPermission(this.project.key, env.name, event.gp).subscribe(() => {
+                        this._toast.success('', this._translate.instant('permission_deleted'));
```
Comment (bnjjj) [inline ui/src/app/views/project/show/project.component.ts]: and if it fails ?

File: ui/src/app/views/project/show/project.component.ts [hunk 4]
```diff
@@ -112,6 +121,43 @@ export class ProjectShowComponent implements OnInit, OnDestroy {
...
+    addEnvPermEvent(event: EnvironmentPermissionEvent, skip?: boolean): void {
+        if (!skip && this.project.externalChange) {
+            this.permEnvWarningModal.show(event);
+        } else {
+            this.permEnvFormLoading = true;
+            this._projectStore.addEnvironmentPermission(this.project.key, event.env.name, event.gp).subscribe(() => {
+                this._toast.success('', this._translate.instant('permission_added'));
+                this.permEnvFormLoading = false;
+            }, () => {
+                this.permEnvFormLoading = false;
```
Comment (bnjjj) [inline ui/src/app/views/project/show/project.component.ts]: silent fail ?

Comment (sguiheux) [inline ui/src/app/views/project/show/project.component.ts]: no, there is a toast spawn on http interceptor on all error

File: ui/src/app/views/project/show/project.component.ts (hunk 2 above)
Comment (sguiheux) [inline ui/src/app/views/project/show/project.component.ts]: same than above

File: ui/src/app/views/project/show/project.component.ts (hunk 3 above)
Comment (sguiheux) [inline ui/src/app/views/project/show/project.component.ts]: same than above

**Pull request merged**